sdks/python/apache_beam/testing/load_tests/par_do_test.py

```diff
@@ -0,0 +1,146 @@
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+"""
+To run test on DirectRunner
+
+python setup.py nosetests \
+    --test-pipeline-options="
+    --iterations=1000
+    --input_options='{
+    \"num_records\": 300,
+    \"key_size\": 5,
+    \"value_size\":15,
+    \"bundle_size_distribution_type\": \"const\",
+    \"bundle_size_distribution_param\": 1,
+    \"force_initial_num_bundles\": 0
+    }'" \
+    --tests apache_beam.testing.load_tests.par_do_test
+
+To run test on other runner (ex. Dataflow):
+
+python setup.py nosetests \
+    --test-pipeline-options="
+        --runner=TestDataflowRunner
+        --project=...
+        --staging_location=gs://...
+        --temp_location=gs://...
+        --sdk_location=./dist/apache-beam-x.x.x.dev0.tar.gz
+        --output=gc
+        --iterations=1000
+        --input_options='{
+        \"num_records\": 1000,
+        \"key_size\": 5,
+        \"value_size\":15,
+        \"bundle_size_distribution_type\": \"const\",
+        \"bundle_size_distribution_param\": 1,
+        \"force_initial_num_bundles\": 0
+        }'" \
+    --tests apache_beam.testing.load_tests.par_do_test
+
+"""
+
+from __future__ import absolute_import
+
+import json
+import logging
+import time
+import unittest
+
+import apache_beam as beam
+from apache_beam.metrics import Metrics
+from apache_beam.testing import synthetic_pipeline
+from apache_beam.testing.test_pipeline import TestPipeline
+
+
+class ParDoTest(unittest.TestCase):
+  def parseTestPipelineOptions(self):
+    return {'numRecords': self.inputOptions.get('num_records'),
+            'keySizeBytes': self.inputOptions.get('key_size'),
+            'valueSizeBytes': self.inputOptions.get('value_size'),
+            'bundleSizeDistribution': {
+                'type': self.inputOptions.get(
+                    'bundle_size_distribution_type', 'const'
+                ),
+                'param': self.inputOptions.get(
+                    'bundle_size_distribution_param', 0
+                )
+            },
+            'forceNumInitialBundles': self.inputOptions.get(
+                'force_initial_num_bundles', 0
+            )
+           }
+
+  def setUp(self):
+    self.pipeline = TestPipeline(is_integration_test=True)
+    self.output = self.pipeline.get_option('output')
+    self.iterations = self.pipeline.get_option('iterations')
+    self.inputOptions = json.loads(self.pipeline.get_option('input_options'))
+
+  class _MeasureTime(beam.DoFn):
```

Comment:
You may be able to check on start-bundle and finish-bundle, to avoid checking too often:

    class _MeasureTime(beam.DoFn):
      def start_bundle(self):
        self.distribution.update(time.time())

      def finish_bundle(self):
        self.distribution.update(time.time())

what do you think?

Reply:
This is a really great idea, thanks a lot!